### PR TITLE
[WIP] Integrate Codecov

### LINKED ISSUE #1638: Add CodeCov
See #1479
We are seeing errors while running the code coverage in development. To see the information for every branch in ci we can add Codecov (https://codecov.io) to get a report on every branch. This will give the information of how much of the added code is tested and not tested.
See also [this](https://github.com/dwyl/repo-badges) on why codecov will be good to add.

Steps to add Codecov to the github repo:
- Login using your github account through https://codecov.io
- select elife-xpub repo from the repos
- Codecov is approved by @giorgiosironi so you should be able to see the settings where you can find the generated token.
- The token needs to be added to the ci pipeline ( this step to be defined by @giorgiosironi )
- Investigate further if the Jenkinsfile should have been updated to include the codecov configuration

[Codecov docs](https://docs.codecov.io/docs)

### PR DESCRIPTION
**Pull request by giorgiosironi**
Closes #1638

### CODE REVIEW

**Comment by giorgiosironi** [comment]
From the build:
```
vault.sh kv get -field token secret/containers/elife-xpub/codecov > .codecov-token— Shell Script
[2019-03-06T18:07:40.255Z] + vault.sh kv get -field token secret/containers/elife-xpub/codecov
ls -l .codecov-token— Shell Script
[2019-03-06T18:07:40.551Z] + ls -l .codecov-token
[2019-03-06T18:07:40.551Z] -rw-rw-r-- 1 ubuntu ubuntu 36 Mar  6 18:07 .codecov-token
```
This gets the token into a local file outside of the containers. We should tweak it to pass it to the right script, probably inside the appropriate container.

**Review by Cooryd — APPROVED** [review]

**Review by diversemix — APPROVED** [review]

**Comment by giorgiosironi** [comment]
Not sure if you want to merge this as-is, or if you want to build on top of this branch to pass the token wherever is needed.

**Comment by diversemix** [comment]
Happy for @SemraSh to make that call!

File: Jenkinsfile [hunk 1]
```diff
@@ -24,6 +24,9 @@ elifePipeline {
                 'test': {
                     withCommitStatus({
                         sh "IMAGE_TAG=${commit} docker-compose -f docker-compose.yml -f docker-compose.ci.yml run --rm --name elife-xpub_app_test app npm test"
+                        sh "vault.sh kv get -field token secret/containers/elife-xpub/codecov > .codecov-token"
+                        sh "ls -l .codecov-token"
+                        sh "curl -s https://codecov.collibra.com/bash | bash -s -- -t @.codecov-token"
```
Comment (Cooryd) [inline Jenkinsfile]: I believe this should be ran within the container so should be part of the docker-compose file. probably https://github.com/elifesciences/elife-xpub/blob/develop/docker-compose.ci.yml as it shouldn't be needed in the production build

Comment (giorgiosironi) [reply Jenkinsfile]: Yes, it should run in the containers to provide its own reproducible environment. The file can be mounted with `volumes` in the `docker-compose.ci.yml` you mentioned.

File: Jenkinsfile [hunk 2]
```diff
@@ -23,7 +23,8 @@ elifePipeline {
                 },
                 'test': {
                     withCommitStatus({
-                        sh "IMAGE_TAG=${commit} docker-compose -f docker-compose.yml -f docker-compose.ci.yml run --rm --name elife-xpub_app_test app npm test"
+                        sh "vault.sh kv get -field token secret/containers/elife-xpub/codecov > .codecov-token"
+                        sh "IMAGE_TAG=${commit} docker-compose -f docker-compose.yml -f docker-compose.ci.yml run -e CODECOV_TOKEN=$(cat .codecov-token) --rm --name elife-xpub_app_test app npm run test-ci"
```
Comment (Cooryd) [inline Jenkinsfile]: `$(...)` should be `${...}`

Comment (Cooryd) [reply Jenkinsfile]: or maybe not 🤷‍♂️

Comment (giorgiosironi) [reply Jenkinsfile]: `${...}` is a (Groovy) substitution of a variable into the string while the Jenkinsfile gets executed.
`$(...)` is a (bash) substitution of a command's result into the string while the command in `sh` gets executed.

**Comment by giorgiosironi** [comment]
```
[2019-03-11T16:17:50.035Z]  / ____|        | |
[2019-03-11T16:17:50.035Z] | |     ___   __| | ___  ___ _____   __
[2019-03-11T16:17:50.035Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2019-03-11T16:17:50.035Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2019-03-11T16:17:50.035Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2019-03-11T16:17:50.035Z]                               Bash-8a28df4
[2019-03-11T16:17:50.035Z] 
[2019-03-11T16:17:50.035Z] 
[2019-03-11T16:17:50.035Z] x> No CI provider detected.
[2019-03-11T16:17:50.035Z]     Testing inside Docker? http://docs.codecov.io/docs/testing-with-docker
[2019-03-11T16:17:50.035Z]     Testing with Tox? https://docs.codecov.io/docs/python#section-testing-with-tox
[2019-03-11T16:17:50.035Z]     project root: .
[2019-03-11T16:17:50.035Z] --> token set from env
[2019-03-11T16:17:50.035Z] not found!
[2019-03-11T16:17:50.035Z]     Yaml not found, that's ok! Learn more at http://docs.codecov.io/docs/codecov-yaml
[2019-03-11T16:17:50.035Z] ==> Running gcov in . (disable via -X gcov)
[2019-03-11T16:17:50.968Z] ==> Python coveragepy not found
[2019-03-11T16:17:50.968Z] ==> Searching for coverage reports in:
[2019-03-11T16:17:50.968Z]     + .
[2019-03-11T16:17:51.226Z]     -> Found 3 reports
[2019-03-11T16:17:51.226Z] ==> Detecting git/mercurial file structure
[2019-03-11T16:17:53.123Z] ==> Reading reports
[2019-03-11T16:17:53.123Z]     + ./coverage/lcov.info bytes=85617
[2019-03-11T16:17:53.123Z]     + ./coverage/clover.xml bytes=198882
[2019-03-11T16:17:53.123Z]     + ./coverage/coverage-final.json bytes=482622
[2019-03-11T16:17:53.123Z] ==> Appending adjustments
[2019-03-11T16:17:53.123Z]     http://docs.codecov.io/docs/fixing-reports
[2019-03-11T16:17:55.648Z]     + Found adjustments
[2019-03-11T16:17:55.648Z] ==> Gzipping contents
[2019-03-11T16:17:55.905Z] ==> Uploading reports
[2019-03-11T16:17:55.905Z]     url: https://codecov.collibra.com
[2019-03-11T16:17:55.905Z]     query: branch=&commit=&build=&build_url=&name=&tag=&slug=&service=&flags=&pr=&job=
[2019-03-11T16:17:55.905Z]     -> Pinging Codecov
[2019-03-11T16:17:55.905Z] https://codecov.collibra.com/upload/v4?package=bash-8a28df4&token=99f90f27-3906-4979-98b8-0f51bee6898c&branch=&commit=&build=&build_url=&name=&tag=&slug=&service=&flags=&pr=&job=
[2019-03-11T16:17:56.472Z]     -> Uploading
[2019-03-11T16:17:56.472Z]     X> Failed to upload
[2019-03-11T16:17:56.472Z]     -> Sleeping for 30s and trying again...
[2019-03-11T16:18:53.147Z]     -> Pinging Codecov
[2019-03-11T16:18:53.147Z] https://codecov.collibra.com/upload/v4?package=bash-8a28df4&token=99f90f27-3906-4979-98b8-0f51bee6898c&branch=&commit=&build=&build_url=&name=&tag=&slug=&service=&flags=&pr=&job=
[2019-03-11T16:18:53.147Z]     -> Uploading
[2019-03-11T16:18:53.148Z]     X> Failed to upload
[2019-03-11T16:18:53.148Z]     -> Sleeping for 30s and trying again...
[2019-03-11T16:18:57.338Z]     -> Pinging Codecov
[2019-03-11T16:18:57.338Z] https://codecov.collibra.com/upload/v4?package=bash-8a28df4&token=99f90f27-3906-4979-98b8-0f51bee6898c&branch=&commit=&build=&build_url=&name=&tag=&slug=&service=&flags=&pr=&job=
[2019-03-11T16:18:57.338Z]     -> Uploading
[2019-03-11T16:18:57.338Z]     X> Failed to upload
[2019-03-11T16:18:57.338Z]     -> Sleeping for 30s and trying again...
[2019-03-11T16:19:29.390Z]     -> Pinging Codecov
[2019-03-11T16:19:29.390Z] https://codecov.collibra.com/upload/v4?package=bash-8a28df4&token=99f90f27-3906-4979-98b8-0f51bee6898c&branch=&commit=&build=&build_url=&name=&tag=&slug=&service=&flags=&pr=&job=
[2019-03-11T16:19:29.390Z]     -> Uploading
[2019-03-11T16:19:29.390Z]     X> Failed to upload
[2019-03-11T16:19:29.390Z]     -> Sleeping for 30s and trying again...
[2019-03-11T16:20:01.459Z]     -> Uploading to Codecov
[2019-03-11T16:20:01.459Z]     HTTP 400
[2019-03-11T16:20:01.459Z] missing required properties: [&#39;commit&#39;]
```
It's probably not able to find the commit because there is no Git repository inside the Docker image. We can pass in the `commit` variable from the Jenkins file if there is a way to specify the value ourselves.

**Comment by giorgiosironi** [comment]
Checklist item: if the tests fail, verify that the build still fails; not sure what happens with the concatenated coverage command and how it affects the exit code.

**Pull request closed without merging**